[UnifiedPDF] [iOS] Zoom locked after using per site zoom

aprotyas opened the pull request:
#### cb86c682ac0167a7161e57f5742490f9650e3ec3
<pre>
[UnifiedPDF] [iOS] Zoom locked after using per site zoom
<a href="https://bugs.webkit.org/show_bug.cgi?id=287857">https://bugs.webkit.org/show_bug.cgi?id=287857</a>
<a href="https://rdar.apple.com/144613344">rdar://144613344</a>

Reviewed by Wenson Hsieh.

Currently, Safari&apos;s page format menu and the pinch gesture zooming do
not play hand in hand, manifesting in unexpected symptoms such as the
bug title.

This patch serves a harder whammy placed on our ability to respect page
scale adjustments driven through Safari&apos;s page format menu. We do so by
dropping scale updates on the floor when we detect a main frame plugin
which does not scale factor updates. We also take care to restore the
client-driven layout size scale factor to 1 in place of some other
default. (again, to prevent the &quot;locking&quot; seen above)

API test: UnifiedPDF.ShouldNotRespectSetViewScale

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setViewScale:]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::mainFramePluginOverridesViewScale const):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST):

Canonical link: <a href="https://commits.webkit.org/290565@main">https://commits.webkit.org/290565@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c160b87e0368d41a8b243807185c6b17e2a2d4a5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90441 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9973 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45376 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95451 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41223 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92493 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10364 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18292 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/69623 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27194 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93442 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7935 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82041 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49972 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/7657 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/36394 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40353 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/77992 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37477 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97281 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17633 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12963 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/78611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17891 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77861 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77831 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19219 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22285 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20893 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/10955 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17643 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/22975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17382 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20837 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19166 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -1810,6 +1810,7 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     void willStartUserTriggeredZooming();
     void didEndUserTriggeredZooming();
     bool mainFramePluginHandlesPageScaleGesture() const { return m_mainFramePluginHandlesPageScaleGesture; }
+    bool hasMainFramePlugin() const { return m_pluginMaxZoomFactor.has_value() || m_pluginMinZoomFactor.has_value(); }
```
Hm... it seems a bit unfortunate to tie `hasMainFramePlugin()` to zooming state, though it seems correct in this context. Maybe we could clarify this by renaming `hasMainFramePlugin()` to something more specific, like `mainFramePluginOverridesViewScale()`, and do:

```
    if (_page->mainFramePluginOverridesViewScale()) {
        if (_page->layoutSizeScaleFactorFromClient() != 1)
            _page->setViewportConfigurationViewLayoutSize(_page->viewLayoutSize(), 1, _page->minimumEffectiveDeviceWidth());
        return;
    }
```

…

```
bool WebPageProxy::mainFramePluginOverridesViewScale() const
{
    if (m_mainFramePluginHandlesPageScaleGesture)
        return false;
    return m_pluginMaxZoomFactor.has_value() || m_pluginMinZoomFactor.has_value();
}
```

aprotyas replied:
Yes, I agree, it seems weird for `hasMainFramePlugin()` to show up in zooming logic.

aprotyas replied:
I will go with your suggestion

aprotyas replied:
Changed in https://github.com/WebKit/WebKit/commit/c160b87e0368d41a8b243807185c6b17e2a2d4a5

aprotyas committed:
[UnifiedPDF] [iOS] Zoom locked after using per site zoom
https://bugs.webkit.org/show_bug.cgi?id=287857
rdar://144613344

Reviewed by Wenson Hsieh.

Currently, Safari's page format menu and the pinch gesture zooming do
not play hand in hand, manifesting in unexpected symptoms such as the
bug title.

This patch serves a harder whammy placed on our ability to respect page
scale adjustments driven through Safari's page format menu. We do so by
dropping scale updates on the floor when we detect a main frame plugin
which does not scale factor updates. We also take care to restore the
client-driven layout size scale factor to 1 in place of some other
default. (again, to prevent the "locking" seen above)

API test: UnifiedPDF.ShouldNotRespectSetViewScale

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setViewScale:]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::mainFramePluginOverridesViewScale const):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST):

Canonical link: https://commits.webkit.org/290565@main
